File: shared/src/commonMain/kotlin/com/nailtryon/domain/ConnectedComponents.kt
package com.nailtryon.domain

/**
 * Pure Kotlin implementation of Connected Components Labeling (CCL)
 * using two-pass algorithm with Union-Find for nail segmentation mask processing.
 *
 * This is shared between Android (TFLite) and iOS (CoreML) platforms.
 */
object ConnectedComponents {

    data class ComponentBox(
        var left: Int = Int.MAX_VALUE,
        var top: Int = Int.MAX_VALUE,
        var right: Int = Int.MIN_VALUE,
        var bottom: Int = Int.MIN_VALUE
    ) {
        fun width() = right - left + 1
        fun height() = bottom - top + 1
        fun addPoint(x: Int, y: Int) {
            if (x < left) left = x
            if (x > right) right = x
            if (y < top) top = y
            if (y > bottom) bottom = y
        }
    }

    data class ComponentProperties(
        val centerX: Double,
        val centerY: Double,
        val boxWidth: Int,
        val boxHeight: Int
    )

    /**
     * Find connected components in a binary segmentation mask.
     * @param confidenceValues Float array from model output (size = inputSize * inputSize)
     * @param inputSize Width/height of the square input
     * @param threshold Confidence threshold (default 0.8)
     * @return Pair of (labels array, bounding boxes map)
     */
    fun findComponents(
        confidenceValues: FloatArray,
        inputSize: Int,
        threshold: Float = 0.8f
    ): Pair<IntArray, Map<Int, ComponentBox>> {
        val labels = IntArray(inputSize * inputSize)
        var nextLabel = 1
        val parent = IntArray(inputSize * inputSize / 2 + 1) { it }

        fun find(i: Int): Int {
            var root = i
            while (root != parent[root]) root = parent[root]
            var curr = i
            while (curr != root) {
                val next = parent[curr]
                parent[curr] = root
                curr = next
            }
            return root
        }

        fun union(i: Int, j: Int) {
            val rootI = find(i)
            val rootJ = find(j)
            if (rootI != rootJ) parent[rootJ] = rootI
        }

        // First pass: assign labels
        for (y in 0 until inputSize) {
            for (x in 0 until inputSize) {
                val index = y * inputSize + x
                if (confidenceValues[index] > threshold) {
                    var leftLabel = 0
                    var topLabel = 0
                    if (x > 0) leftLabel = labels[index - 1]
                    if (y > 0) topLabel = labels[index - inputSize]

                    labels[index] = when {
                        leftLabel != 0 && topLabel != 0 -> {
                            if (leftLabel != topLabel) union(leftLabel, topLabel)
                            minOf(leftLabel, topLabel)
                        }
                        leftLabel != 0 -> leftLabel
                        topLabel != 0 -> topLabel
                        else -> {
                            val newLabel = nextLabel++
                            newLabel
                        }
                    }
                }
            }
        }

        // Second pass: resolve equivalence classes and compute bounding boxes
        val finalBoundingBoxes = HashMap<Int, ComponentBox>()
        val rootToFinalLabel = HashMap<Int, Int>()
        var finalLabelCounter = 1

        for (i in labels.indices) {
            if (labels[i] > 0) {
                val root = find(labels[i])
                val finalId = rootToFinalLabel.getOrPut(root) { finalLabelCounter++ }
                labels[i] = finalId
                val x = i % inputSize
                val y = i / inputSize
                val box = finalBoundingBoxes.getOrPut(finalId) { ComponentBox() }
                box.addPoint(x, y)
            }
        }

        return Pair(labels, finalBoundingBoxes)
    }

    /**
     * Calculate centroid and bounding box properties for each labeled component.
     */
    fun calculateProperties(
        labels: IntArray,
        boxes: Map<Int, ComponentBox>,
        inputSize: Int
    ): Map<Int, ComponentProperties> {
        val props = HashMap<Int, ComponentProperties>()
        val moments = HashMap<Int, DoubleArray>()

        for (i in labels.indices) {
            val label = labels[i]
            if (label > 0) {
                val y = i / inputSize
                val x = i % inputSize
                val m = moments.getOrPut(label) { DoubleArray(3) }
                m[0]++
                m[1] += x.toDouble()
                m[2] += y.toDouble()
            }
        }

        for ((label, m) in moments) {
            val n = m[0]
            val meanX = m[1] / n
            val meanY = m[2] / n
            val box = boxes[label]
            props[label] = ComponentProperties(
                centerX = meanX,
                centerY = meanY,
                boxWidth = box?.width() ?: 10,
                boxHeight = box?.height() ?: 10
            )
        }

        return props
    }
}

File: iosApp/iosApp/ConnectedComponents.swift
import Foundation
import UIKit

/// Ported from shared Kotlin `ConnectedComponents.kt`
/// Connected component labeling and property calculation for nail segmentation.
class ConnectedComponents {
    
    struct ComponentProperties {
        let centerX: Float
        let centerY: Float
        let boxWidth: Float
        let boxHeight: Float
        let pixelCount: Int
    }
    
    /// Find connected components in the confidence map using thresholding.
    /// - Parameters:
    ///   - confidenceValues: Flat array of confidence values (inputSize * inputSize)
    ///   - inputSize: Dimension of the square input
    ///   - threshold: Confidence threshold (default 0.8)
    /// - Returns: (labels array, bounding boxes dictionary [componentId -> minX, minY, maxX, maxY])
    static func findComponents(
        _ confidenceValues: [Float],
        inputSize: Int,
        threshold: Float = 0.8
    ) -> (labels: [Int], boundingBoxes: [Int: (minX: Int, minY: Int, maxX: Int, maxY: Int)]) {
        
        var labels = [Int](repeating: 0, count: inputSize * inputSize)
        var boundingBoxes: [Int: (minX: Int, minY: Int, maxX: Int, maxY: Int)] = [:]
        var currentLabel = 0
        var equivalences: [Int: Int] = [:]
        
        func find(_ x: Int) -> Int {
            var root = x
            while equivalences[root] != nil && equivalences[root] != root {
                root = equivalences[root]!
            }
            // Path compression
            var cur = x
            while equivalences[cur] != nil && equivalences[cur] != cur {
                let next = equivalences[cur]!
                equivalences[cur] = root
                cur = next
            }
            return root
        }
        
        // First pass: assign labels with connected component analysis
        for y in 0..<inputSize {
            for x in 0..<inputSize {
                let idx = y * inputSize + x
                guard confidenceValues[idx] > threshold else { continue }
                
                let upIdx = (y > 0) ? ((y - 1) * inputSize + x) : -1
                let leftIdx = (x > 0) ? (y * inputSize + (x - 1)) : -1
                let upLabel = (upIdx >= 0 && labels[upIdx] > 0) ? labels[upIdx] : 0
                let leftLabel = (leftIdx >= 0 && labels[leftIdx] > 0) ? labels[leftIdx] : 0
                
                if upLabel == 0 && leftLabel == 0 {
                    currentLabel += 1
                    labels[idx] = currentLabel
                    boundingBoxes[currentLabel] = (x, y, x, y)
                } else if upLabel != 0 && leftLabel == 0 {
                    labels[idx] = upLabel
                    if var box = boundingBoxes[upLabel] {
                        box.maxX = max(box.maxX, x)
                        box.maxY = max(box.maxY, y)
                        boundingBoxes[upLabel] = box
                    }
                } else if upLabel == 0 && leftLabel != 0 {
                    labels[idx] = leftLabel
                    if var box = boundingBoxes[leftLabel] {
                        box.maxX = max(box.maxX, x)
                        box.maxY = max(box.maxY, y)
                        boundingBoxes[leftLabel] = box
                    }
                } else {
                    // Both neighbors have labels - merge
                    let minLabel = min(upLabel, leftLabel)
                    let maxLabel = max(upLabel, leftLabel)
                    labels[idx] = minLabel
                    equivalences[maxLabel] = minLabel
                    
                    if var minBox = boundingBoxes[minLabel],
                       var maxBox = boundingBoxes[maxLabel] {
                        minBox.minX = min(minBox.minX, x, maxBox.minX)
                        minBox.minY = min(minBox.minY, y, maxBox.minY)
                        minBox.maxX = max(minBox.maxX, x, maxBox.maxX)
                        minBox.maxY = max(minBox.maxY, y, maxBox.maxY)
                        boundingBoxes[minLabel] = minBox
                    }
                }
            }
        }
        
        // Second pass: resolve equivalences
        for i in 0..<labels.count {
            if labels[i] > 0 {
                labels[i] = find(labels[i])
            }
        }
        
        // Merge bounding boxes after equivalence resolution
        var mergedBoxes: [Int: (minX: Int, minY: Int, maxX: Int, maxY: Int)] = [:]
        for (label, box) in boundingBoxes {
            let rootLabel = find(label)
            if let existing = mergedBoxes[rootLabel] {
                mergedBoxes[rootLabel] = (
                    min(existing.minX, box.minX),
                    min(existing.minY, box.minY),
                    max(existing.maxX, box.maxX),
                    max(existing.maxY, box.maxY)
                )
            } else {
                mergedBoxes[rootLabel] = box
            }
        }
        
        return (labels, mergedBoxes)
    }
    
    /// Calculate properties for each connected component.
    static func calculateProperties(
        labels: [Int],
        boundingBoxes: [Int: (minX: Int, minY: Int, maxX: Int, maxY: Int)],
        inputSize: Int
    ) -> [Int: ComponentProperties] {
        
        var properties: [Int: ComponentProperties] = [:]
        
        // Calculate pixel counts
        var pixelCounts: [Int: Int] = [:]
        for label in labels {
            if label > 0 {
                pixelCounts[label] = (pixelCounts[label] ?? 0) + 1
            }
        }
        
        for (label, box) in boundingBoxes {
            // Filter out noise: require minimum pixel count
            let count = pixelCounts[label] ?? 0
            guard count > 20 else { continue }
            
            let centerX = Float(box.minX + box.maxX) / 2.0
            let centerY = Float(box.minY + box.maxY) / 2.0
            let boxWidth = Float(box.maxX - box.minX)
            let boxHeight = Float(box.maxY - box.minY)
            
            properties[label] = ComponentProperties(
                centerX: centerX,
                centerY: centerY,
                boxWidth: boxWidth,
                boxHeight: boxHeight,
                pixelCount: count
            )
        }
        
        return properties
    }
}
